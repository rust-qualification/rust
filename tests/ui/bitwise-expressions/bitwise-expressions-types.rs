// Bitwise expressions: 58fa3a4b-f15b-41f3-b8b3-2f607b8b8688

pub fn main() {
    let a: bool = true;
    let b: bool = false;
    
    let a_or_b = a | b;
    let a_and_b = a & b;
    let a_xor_b = a ^ b;
    let not_b = !b;
    let a_shift = a << b; 
    //~^ ERROR no implementation for `bool << bool`
    let b_shift = a >> b; 
    //~^ ERROR no implementation for `bool >> bool`

    let c: i8 = 0b1010;
    let d: i8 = 0b0101;
    
    let c_or_d = c | d;
    let c_and_d = c & d;
    let c_xor_d = c ^ d;
    let not_d = !d;
    let c_shift = c << d; 
    let d_shift = c >> d; 
    
    let e: i8 = 3;
    let f: i8 = 4;
    
    let e_or_f = e | f;
    let e_and_f = e & f;
    let e_xor_f = e ^ f;
    let not_f = !f;
    let e_shift = e << f; 
    let f_shift = e >> f; 

    // invalid types
    let g: f32 = 1.2;
    let h: f32 = 1.5;
    
    let g_or_h = g | h; 
    //~^ ERROR no implementation for `f32 | f32`
    let g_and_h = g & h;
    //~^ ERROR no implementation for `f32 & f32`
    let g_xor_h = g ^ h;
    //~^ ERROR no implementation for `f32 ^ f32`
    let not_h = !h;
    //~^ ERROR cannot apply unary operator `!` to type `f32`
    let g_shift = g << h; 
    //~^ ERROR no implementation for `f32 << f32`
    let h_shift = g >> h; 
    //~^ ERROR no implementation for `f32 >> f32`
    
    // invalid types
    let i: char = 'a';
    let j: char = 'b';
    
    let i_or_j = i | j;
    //~^ ERROR no implementation for `char | char`
    let i_and_j = i & j;
    //~^ ERROR no implementation for `char & char`
    let i_xor_j = i ^ j;
    //~^ ERROR no implementation for `char ^ char`
    let not_j = !j;
    //~^ ERROR cannot apply unary operator `!` to type `char`
    let i_shift = i << j; 
    //~^ ERROR no implementation for `char << char`
    let j_shift = i >> j; 
    //~^ ERROR no implementation for `char >> char`
    

    // invalid types
    let k:  String = String::from("Hello");
    let l:  String = String::from("world");

    let k_or_l = k | l;
    //~^ ERROR no implementation for `String | String`
    let k_and_l = k & l;
    //~^ ERROR no implementation for `String & String`
    let k_xor_l = k ^ l;
    //~^ ERROR no implementation for `String ^ String`
    let not_l = !l;   
    //~^ ERROR cannot apply unary operator `!` to type `String`
    let k_shift = k << l; 
    //~^ ERROR no implementation for `String << String`
    let l_shift = k >> l; 
    //~^ ERROR no implementation for `String >> String`
    

    
    // invalid types
    let m: i8 = 3;
    let n: bool = false;
    
    let m_or_n = m | n;
    //~^ ERROR no implementation for `i8 | bool`
    let m_and_n = m & n;
    //~^ ERROR no implementation for `i8 & bool`
    let m_xor_n = m ^ n;
    //~^ ERROR no implementation for `i8 ^ bool`
    let m_shift = m << n; 
    //~^ ERROR no implementation for `i8 << bool`
    let n_shift = m >> n; 
    //~^ ERROR no implementation for `i8 >> bool`
    
}