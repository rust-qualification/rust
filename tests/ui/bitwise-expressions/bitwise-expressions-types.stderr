error[E0369]: no implementation for `f32 | f32`
  --> $DIR/bitwise-expressions-types.rs:33:20
   |
LL |     let g_or_h = g | h;
   |                  - ^ - f32
   |                  |
   |                  f32

error[E0369]: no implementation for `f32 & f32`
  --> $DIR/bitwise-expressions-types.rs:34:21
   |
LL |     let g_and_h = g & h;
   |                   - ^ - f32
   |                   |
   |                   f32

error[E0369]: no implementation for `f32 ^ f32`
  --> $DIR/bitwise-expressions-types.rs:35:21
   |
LL |     let g_xor_h = g ^ h;
   |                   - ^ - f32
   |                   |
   |                   f32

error[E0600]: cannot apply unary operator `!` to type `f32`
  --> $DIR/bitwise-expressions-types.rs:36:17
   |
LL |     let not_h = !h;
   |                 ^^ cannot apply unary operator `!`

error[E0369]: no implementation for `char | char`
  --> $DIR/bitwise-expressions-types.rs:43:20
   |
LL |     let i_or_j = i | j;
   |                  - ^ - char
   |                  |
   |                  char

error[E0369]: no implementation for `char & char`
  --> $DIR/bitwise-expressions-types.rs:44:21
   |
LL |     let i_and_j = i & j;
   |                   - ^ - char
   |                   |
   |                   char

error[E0369]: no implementation for `char ^ char`
  --> $DIR/bitwise-expressions-types.rs:45:21
   |
LL |     let i_xor_j = i ^ j;
   |                   - ^ - char
   |                   |
   |                   char

error[E0600]: cannot apply unary operator `!` to type `char`
  --> $DIR/bitwise-expressions-types.rs:46:17
   |
LL |     let not_j = !j;
   |                 ^^ cannot apply unary operator `!`

error[E0369]: no implementation for `String | String`
  --> $DIR/bitwise-expressions-types.rs:54:20
   |
LL |     let k_or_l = k | l;
   |                  - ^ - String
   |                  |
   |                  String

error[E0369]: no implementation for `String & String`
  --> $DIR/bitwise-expressions-types.rs:55:21
   |
LL |     let k_and_l = k & l;
   |                   - ^ - String
   |                   |
   |                   String

error[E0369]: no implementation for `String ^ String`
  --> $DIR/bitwise-expressions-types.rs:56:21
   |
LL |     let k_xor_l = k ^ l;
   |                   - ^ - String
   |                   |
   |                   String

error[E0600]: cannot apply unary operator `!` to type `String`
  --> $DIR/bitwise-expressions-types.rs:57:17
   |
LL |     let not_l = !l;   
   |                 ^^ cannot apply unary operator `!`

error[E0277]: no implementation for `i8 | bool`
  --> $DIR/bitwise-expressions-types.rs:64:20
   |
LL |     let m_or_n = m | n;
   |                    ^ no implementation for `i8 | bool`
   |
   = help: the trait `BitOr<bool>` is not implemented for `i8`
   = help: the following other types implement trait `BitOr<Rhs>`:
             `&'a i8` implements `BitOr<i8>`
             `&i8` implements `BitOr<&i8>`
             `i8` implements `BitOr<&i8>`
             `i8` implements `BitOr`

error[E0277]: no implementation for `i8 & bool`
  --> $DIR/bitwise-expressions-types.rs:65:21
   |
LL |     let m_and_n = m & n;
   |                     ^ no implementation for `i8 & bool`
   |
   = help: the trait `BitAnd<bool>` is not implemented for `i8`
   = help: the following other types implement trait `BitAnd<Rhs>`:
             `&'a i8` implements `BitAnd<i8>`
             `&i8` implements `BitAnd<&i8>`
             `i8` implements `BitAnd<&i8>`
             `i8` implements `BitAnd`

error[E0277]: no implementation for `i8 ^ bool`
  --> $DIR/bitwise-expressions-types.rs:66:21
   |
LL |     let m_xor_n = m ^ n;
   |                     ^ no implementation for `i8 ^ bool`
   |
   = help: the trait `BitXor<bool>` is not implemented for `i8`
   = help: the following other types implement trait `BitXor<Rhs>`:
             `&'a i8` implements `BitXor<i8>`
             `&i8` implements `BitXor<&i8>`
             `i8` implements `BitXor<&i8>`
             `i8` implements `BitXor`

error: aborting due to 15 previous errors

Some errors have detailed explanations: E0277, E0369, E0600.
For more information about an error, try `rustc --explain E0277`.
